[Your Voice Matters (YVM)] CARDS-2384: clinic & survey; CARDS-2387: CSV export; CARDS-2433: survey scheduling conflicts; CARDS-2385: clarity import & survey assignment

marta- opened the pull request:
**DRAFT until [CARDS-2449](https://phenotips.atlassian.net/browse/CARDS-2449) (_Update RecentVisitDiscardFilter_) is implemented, so we can properly test survey frequency.**

**Summary of changes:**
* Added sidebar entry for YVM, email templates, email schedule ([CARDS-2384](https://phenotips.atlassian.net/browse/CARDS-2384))
* Configured the YVM schedule and removed the conflictMode:any setting from all other clinics ([CARDS-2433](https://phenotips.atlassian.net/browse/CARDS-2433));  going forward, conflicts between these other clinics should be handled at import via  [CARDS-2449](https://phenotips.atlassian.net/browse/CARDS-2449) (_Update RecentVisitDiscardFilter_) - **not yet implemented**
* Added the YVM questionnaire to weekly exports ([CARDS-2387](https://phenotips.atlassian.net/browse/CARDS-2387))
* Configured the clarity import & survey assignment for YVM ([CARDS-2385](https://phenotips.atlassian.net/browse/CARDS-2385))

**Testing:**
* Emailing YVM surveys and the appearance of the emails
* YVM patient workflow (filling out the survey)
* YVM survey scheduling (at most one every 6 months)

**Regression testing, to be done AFTER 2449:**
* survey scheduling across Your Experience


[CARDS-2384]: https://phenotips.atlassian.net/browse/CARDS-2384?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CARDS-2387]: https://phenotips.atlassian.net/browse/CARDS-2387?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[CARDS-2433]: https://phenotips.atlassian.net/browse/CARDS-2433?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[CARDS-2449]: https://phenotips.atlassian.net/browse/CARDS-2449?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[CARDS-2385]: https://phenotips.atlassian.net/browse/CARDS-2385?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

marta- committed:
CARDS-2384: Configure the YVM clinic

marta- committed:
CARDS-2387: Configure CSV export for YVM

sdumitriu commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/apps/cards/clinics/PHM-YVM/mailTemplates/InitialInvitation/ROOT.json:
```diff
@@ -0,0 +1,8 @@
+  "jcr:primaryType": "cards:EmailTemplate",
+  "senderAddress": "your.voice.matters@uhn.ca",
+  "senderName": "Your Voice Matters",
+  "subject": "Your Voice Matters: We are listening",
```
Oh, that sounds creepy, but maybe it's just me coming from a place where phone lines were often spied on.

marta- replied:
I agree, but that's the wording that came from Ontario health.

sdumitriu commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/apps/cards/clinics/PHM-YVM/mailTemplates/InitialInvitation/bodyTemplate.html:
```diff
@@ -0,0 +1,45 @@
+    <p>Dear ${first_name} ${last_name},</p>
+    <p>
+      <strong>Your Voice Matters</strong> is a set of questions that gives you the chance to share your cancer care
+      experience. Your answers will help us know how to improve the experience for patients and care
```
"Share your cancer care experience"? Is it just me who finds this phrasing confusing?

marta- replied:
I agree, but that's the wording that came from Ontario health.

sdumitriu commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/apps/cards/clinics/PHM-YVM/mailTemplates/InitialInvitation/bodyTemplate.html:
```diff
@@ -0,0 +1,45 @@
+      cancer care appointment. It will take <strong>about 5 minutes</strong> to complete this survey.
+      This survey is <strong>confidential</strong> (private).
+      Your cancer care team will not see your responses to the questions.
+      Your answers are combined with other people’s responses to help Ontario Health (Cancer Care Ontario)
```
I would like to see a link to "Cancer Care Ontario" here.

marta- replied:
Good idea

marta- replied:
Done in 9d2524e

sdumitriu commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/apps/cards/clinics/PHM-YVM/mailTemplates/InitialInvitation/bodyTemplate.html:
```diff
@@ -0,0 +1,45 @@
+      </p>
+    </div>
+    <p>
+      If you want to provide detailed feedback about your experience directly to the hospital, please contact your.voice.matters@uhn.ca.
```
mailto?

marta- replied:
Good catch

marta- replied:
Done in 9d2524e

sdumitriu commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/apps/cards/clinics/PHM-YVM/mailTemplates/InitialInvitation/bodyTemplate.html:
```diff
@@ -0,0 +1,45 @@
+      If you want to provide detailed feedback about your experience directly to the hospital, please contact your.voice.matters@uhn.ca.
+    </p>
+    <p>
+      If you have any questions, concerns, or wish to be removed from the survey, please contact us at yourexperience@uhn.ca.
```
Is this yourexperience mention intentional or a leftover?

marta- replied:
It's ... preliminary? We're not sure who is supposed to "sign" the email, will find out at the next call.

marta- replied:
> It's ... preliminary? We're not sure who is supposed to "sign" the email and manage complaints, will find out at the next call.

marta- replied:
Updated the contact email for feedback other than unsubscribing  in 53175a5.

sdumitriu commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/apps/cards/clinics/PHM-YVM/mailTemplates/InitialInvitation/bodyTemplate.html:
```diff
@@ -0,0 +1,45 @@
+    </p>
+    <p>
+      If you have any questions, concerns, or wish to be removed from the survey, please contact us at yourexperience@uhn.ca.
+      You can also <a href="${unsubscribeLink}">unsubscribe</a> from all further emails from <strong>Your Experience</strong>.
```
YE here too.

marta- replied:
YE is to be contacted for unsubscribing until further notice.

sdumitriu commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/apps/cards/clinics/PHM-YVM/mailTemplates/InitialInvitation/bodyTemplate.html:
```diff
@@ -0,0 +1,45 @@
+
+    <p>
+      Best regards,<br/>
+      <span class="signature">The Patient Experience Team</span><br/>
```
Here too.

marta- replied:
"The Patient Experience Team" stays, the next line was updated to specify the hospital  in 53175a5.

sdumitriu commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/apps/cards/clinics/PHM-YVM/mailTemplates/ReminderNotification/bodyTemplate.html:
```diff
@@ -0,0 +1,45 @@
+    <p>
+      <strong>Your Voice Matters</strong> is a set of questions that gives you the chance to share your cancer care
+      experience. Your answers will help us know how to improve the experience for patients and care
+      partners at your hospital and across Ontario.
```
I see that these are identical for now, will a "we have not heard back from you" be added?

marta- replied:
Will find out at the next call, the reminder text is temporary.

marta- replied:
Addressed in 53175a5, still subject to change.

sdumitriu commented on prems-resources/feature/src/main/features/feature.json:
```diff
@@ -295,13 +295,39 @@
+      "metricName": "{020} PHM-YVM 1st Reminder Emails Sent",
+      "clinicId": "/Survey/ClinicMapping/258217454",
+      "emailConfiguration": "/apps/cards/clinics/PMH-YVM/mailTemplates/ReminderNotification",
+      "daysToVisit": -4
```
The rule about not sending more than one email in a week doesn't apply here?

marta- replied:
Apparently not.

sdumitriu commented on prems-resources/feature/src/main/features/feature.json:
```diff
@@ -295,13 +295,39 @@
       "frequency.in.days": 7,
       "export.schedule": "0 0 2 ? * SAT *",
       "questionnaires.to.be.exported": [
+        "/Questionnaires/YVM",
```
Shouldn't these be separate from YE exports?

marta- replied:
Not as far as we know as of now. The same team that prepares reports for CIHI will send this data to ISAAC as well .

sdumitriu reviewed (commented):
Do we want a new stats table for YVM?

marta- commented:
> Do we want a new stats table for YVM?

Yes, good catch.

marta- committed:
CARDS-2384: Configure the YVM clinic

Addressing review comments:
* Added mailto: links to emaill addresses in email templates
* Added link to Cancer Care Ontario
* Added entry on the Stats dashboard for YVM

sashaandjic commented:
I am not getting any mails sent. Steps:
-in `compose-cluster`, run 
`python3 generate_compose_yaml.py --dev_docker_image --oak_filesystem --cards_project cards4prems --smtps --smtps_test_container --smtps_test_mail_path ~/cardsmail --composum`
- edit docker-compose.yml file, and at the end of the environment section of the cardsinitial service, add:
   ` - NIGHTLY_NOTIFICATIONS_SCHEDULE=0 * * * * ? *`
- docker-compose build && docker-compose up
- create a new patient information, and visit informations 7, 14, 21 days ago
- wait one minute, then check the contents of the ~/cardsmail folder
- copy/paste template text to: https://msgeml.com/
- stop the instance
- docker-compose down && docker-compose rm && docker volume prune -f && ./cleanup.sh

marta- committed:
CARDS-2384: Configure the YVM clinic
Fixed typo in folder name

marta- committed:
CARDS-2384: Configure the YVM clinic
Make the "close this page" instructions on the summary screen stand out better.

marta- commented:
> I am not getting any mails generated. Steps:

Fixed in 0af5f38 .

sashaandjic reviewed (approved):
tested
![image](https://github.com/data-team-uhn/cards/assets/23243244/1b1dbb51-519b-40bb-9ded-7b9c62c25bdb)

marta- commented on prems-resources/feature/src/main/features/feature.json:
```diff
@@ -295,13 +295,39 @@
+      "metricName": "{020} PHM-YVM 1st Reminder Emails Sent",
+      "clinicId": "/Survey/ClinicMapping/258217454",
+      "emailConfiguration": "/apps/cards/clinics/PMH-YVM/mailTemplates/ReminderNotification",
+      "daysToVisit": -4
```
```suggestion
      "daysToVisit": -5
```

marta- replied:
Done in 97aa620 .

marta- commented on prems-resources/feature/src/main/features/feature.json:
```diff
@@ -295,13 +295,39 @@
+      "clinicId": "/Survey/ClinicMapping/258217454",
+      "emailConfiguration": "/apps/cards/clinics/PMH-YVM/mailTemplates/ReminderNotification",
+      "daysToVisit": -6
+    },
```
Remove the second reminder

marta- replied:
Done in 97aa620 .

marta- commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/Survey/YVM.xml:
```diff
@@ -45,13 +45,14 @@ understand how to best support patients.
+**Your answers have been received. You can now close this page.**
         </value>
         <type>String</type>
     </property>
```
line 79: every 26 weeks overall (including all other YE surveys).

marta- replied:
Done in 216c2ea .

marta- commented on prems-resources/feature/src/main/features/feature.json:
```diff
@@ -318,6 +344,7 @@
       "frequency.in.days": 7,
       "export.schedule": "0 0 2 ? * SAT *",
       "questionnaires.to.be.exported": [
+        "/Questionnaires/YVM",
```
An "in person feedback" section will be added at the end of the YVM survey (similar to the IC survey), which needs to be excluded from export this export and reported separately.

marta- replied:
The new structure of the survey no longer requires any exclusions from export.

marta- committed:
CARDS-2384: Configure the YVM clinic
Updated the email schedule per new specs

marta- committed:
CARDS-2384: Configure the YVM clinic
Updated the survey frequency per new specs

marta- committed:
CARDS-2384: Configure the YVM clinic
Updated the email sender, subject, and text per new specs

marta- committed:
CARDS-2384: Configure the YVM clinic
* Update the contact email in the survey ending message
* Change survey frequency to every 13 weeks (~3 months)

marta- committed:
CARDS-2384: Configure the YVM clinic
Remove the Patient Experience logo from the YVM email footer

sashaandjic commented:
Reminders are being generated after 4 and 6 days.

sashaandjic reviewed (approved):
tested

marta- committed:
CARDS-2384: Configure the YVM clinic
Survey updates

marta- committed:
CARDS-2384: Configure the YVM clinic
Updated the token expiry delay for YVM

marta- committed:
CARDS-2384: Configure the YVM clinic
Email text update, removing "unsubscribe" paragraph from the email text

marta- committed:
CARDS-2384: Configure the YVM clinic
Survey updates

marta- committed:
CARDS-2433: Configure the YVM survey scheduling conflicts
Upsated the YVM scheduling config

marta- committed:
CARDS-2433: Configure the YVM survey scheduling conflicts
Removed `conflictMode: any` from all clinics to allow YVM to be indepenent from everything else.

The conflicts between non-YVM clinics will be handled when importing, by a revamped RecentVisitDiscardFilter.

sdumitriu commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/Extensions/Sidebar/PMH-YVM.json:
```diff
@@ -0,0 +1,8 @@
+  "jcr:primaryType": "cards:Extension",
+  "cards:extensionPointId": "cards/coreUI/sidebar/entry",
+  "cards:extensionName": "PMH YVM",
+  "cards:targetURL": "/content.html/Dashboard/258217454",
```
I really dislike these numbers. Can we just use a name instead?

sdumitriu replied:
Done in 285fa9491.

sdumitriu commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/Questionnaires/YVM.xml:
```diff
@@ -52,7 +52,7 @@
 	</property>
 	<node>
-		<name>s_1</name>
+		<name>section_appt</name>
```
This requires updating the copyanswers config.

sdumitriu replied:
Done in 573d57a3a.

sdumitriu commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/Questionnaires/YVM.xml:
```diff
@@ -413,6 +413,337 @@
+				</property>
+				<property>
+					<name>description</name>
+					<value>Choose all that apply. You can add  other reasons in the text box.</value>
```
```suggestion
					<value>Choose all that apply. You can add other reasons in the text box.</value>
```

sdumitriu commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/Questionnaires/YVM.xml:
```diff
@@ -413,6 +413,337 @@
+				</property>
+				<property>
+					<name>text</name>
+					<value>Did you decline a video virtual visit? If yes, why?</value>
```
I think there should be a `notApplicable` option at the top, `I did not decline a|the virtual visit`.

marta- replied:
Actually let's leave it as is. I remember how important it was to them to learn about barriers. Maybe in addition to requiring in-person care, the patient also doesn't have a private space for a virtual visit. This information is significant and needs to be recorded.

sdumitriu replied:
I don't get your reasoning... How would adding a "no" option to the yes/no answer make the information not be recorded?

marta- replied:
My bad, I misunderstood your suggestion: I thought you wanted to make the **current** first option NA, not add a completely new NA option. I'm ok with that.

sdumitriu commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/Questionnaires/YVM.xml:
```diff
@@ -413,6 +413,337 @@
+					<primaryNodeType>cards:AnswerOption</primaryNodeType>
+					<property>
+						<name>label</name>
+						<value>I do not have access to the personal equipment eg. a computer, tablet, or smart phone</value>
```
This sounds better to me:
```suggestion
						<value>I do not have access to the needed personal equipment, eg. a computer, tablet, or smart phone</value>
```

sdumitriu replied:
Minor issue, now the label and the value are no longer identical. Should they be?

sdumitriu committed:
CARDS-2384: Configure the YVM clinic
Rename the clinic mapping to have a proper name instead of a number

sdumitriu committed:
CARDS-2384: Configure the YVM clinic
Update the copy answer configuration after renaming the sections

marta- committed:
CARDS-2384: Configure the YVM clinic

Minor survey updates from review comments

Co-authored-by: Sergiu Dumitriu <sdumitriu@users.noreply.github.com>

sdumitriu committed:
CARDS-2385: Configure the clarity import & survey assignment for YVM

marta- commented on prems-resources/backend/src/main/java/io/uhndata/cards/prems/internal/importer/DischargedFiller.java:
```diff
@@ -38,7 +38,7 @@ public class DischargedFiller extends AbstractClarityDataProcessor implements Cl
     public DischargedFiller()
     {
-        super(true, new String[] { "prems" }, 30);
+        super(true, new String[] { "prems", "yvm" }, 30);
```
I don't like this

marta- commented on prems-resources/backend/src/main/java/io/uhndata/cards/prems/internal/importer/DischargedFiller.java:
```diff
@@ -38,7 +38,7 @@ public class DischargedFiller extends AbstractClarityDataProcessor implements Cl
     public DischargedFiller()
     {
-        super(true, new String[] { "prems" }, 30);
+        super(true, new String[] { "prems", "yvm" }, 30);
```
I don't like this, we should not have to change the java code to add a new importer.

sdumitriu replied:
I don't like having this many custom processors in each project. In the future they should be turned into a generic configurable processor. To make this one work without changing the java code, we can do one of two things: make it configurable in the featurefile, or remove the supported types altogether so it works all the time. However, I think it's good enough as it is right now, given that it's very custom code for this project only.

marta- replied:
Let's make it configurable in the feature file.

sdumitriu replied:
Replaced with a ConfiguredGenericMapper instance, this doesn't really need to exist.

sdumitriu committed:
CARDS-2385: Configure the clarity import & survey assignment for YVM
Replace the custom DischargeFiller with a configured mapper

sdumitriu committed:
CARDS-2385: Configure the clarity import & survey assignment for YVM
Add a "I did not decline" answer option to the "why did you decline" question

sdumitriu committed:
CARDS-2385: Configure the clarity import & survey assignment for YVM
Add a sample data generation script

sdumitriu committed:
CARDS-2385: Configure the clarity import & survey assignment for YVM
Fixed wrong configuration name

sdumitriu committed:
CARDS-2385: Configure the clarity import & survey assignment for YVM
Configure the SQL table name to use

sdumitriu committed:
CARDS-2385: Configure the clarity import & survey assignment for YVM
Use unique configuration name
